ams/server/src/main/java/com/netease/arctic/server/table/executor/DataExpiringExecutor.java

```diff
@@ -0,0 +1,468 @@
+      }
+      expired = literal.comparator().compare(expireTimestamp, literal.value()) >= 0;
+      if (expired) {
+        partitionFreshness.computeIfPresent(partition,
+            (k, v) -> v.updateExpiredSeq(fileEntry.getSequenceNumber()).incExpiredCount());
+      }
+    }
+
+    return expired;
+  }
+
+  private static boolean willNotRetain(
+      IcebergFileEntry fileEntry,
+      DataExpirationConfig expirationConfig,
+      Map<StructLike, DataFileFreshness> partitionFreshness) {
+    ContentFile<?> contentFile = fileEntry.getFile();
+
+    switch (expirationConfig.expirationLevel) {
+      case PARTITION:
+        // if only partial expired files in a partition, all the files in that partition should be preserved
+        return partitionFreshness.containsKey(contentFile.partition()) &&
+            partitionFreshness.get(contentFile.partition()).expiredDataFileCount ==
+            partitionFreshness.get(contentFile.partition()).totalDataFileCount;
+      case FILE:
+        if (!contentFile.content().equals(FileContent.DATA)) {
+          long seqUpperBound = partitionFreshness.getOrDefault(contentFile.partition(),
+              new DataFileFreshness(Long.MIN_VALUE, Long.MAX_VALUE)).latestExpiredSeq;
+          // only expire delete files with sequence-number less or equal to expired data file
+          // there may be some dangling delete files, they will be cleaned by OrphanFileCleaningExecutor
+          return fileEntry.getSequenceNumber() <= seqUpperBound;
+        } else {
+          return true;
+        }
+      default: return false;
+    }
+  }
+
+  private static Literal<Long> getExpireTimestampLiteral(
+      ContentFile<?> contentFile,
+      Type type,
+      Types.NestedField field,
+      DateTimeFormatter formatter,
+      String numberDateFormatter) {
+    Object upperBound = Conversions.fromByteBuffer(type, contentFile.upperBounds().get(field.fieldId()));
+    Literal<Long> literal = Literal.of(Long.MAX_VALUE);
+    if (null == upperBound) {
+      return literal;
+    } else if (upperBound instanceof Long) {
+      switch (type.typeId()) {
+        case TIMESTAMP:
+          // nanosecond -> millisecond
+          literal = Literal.of((Long) upperBound / 1000);
+          break;
+        default:
+          if (numberDateFormatter.equals(EXPIRE_TIMESTAMP_MS)) {
+            literal = Literal.of((Long) upperBound);
+          } else if (numberDateFormatter.equals(EXPIRE_TIMESTAMP_S)) {
+            // second -> millisecond
+            literal = Literal.of((Long) upperBound * 1000);
+          }
+      }
+    } else if (type.typeId().equals(Type.TypeID.STRING)) {
+      literal = Literal.of(
+          LocalDate.parse(upperBound.toString(), formatter)
+              .atStartOfDay()
+              .atZone(getDefaultZoneId(field)).toInstant().toEpochMilli());
+    }
+    return literal;
+  }
+
+  protected static class DataExpirationConfig {
```

Comment:
I think it would be better to split it because `ExpiringDataConfig` will be persisted in the database and `DataExpirationConfig` has multiple fields that are not easy to serialize.

Reply:
Why ExpiringDataConfig will   be persisted in the database

Reply:
```
void updateTableRuntime(@Param("runtime") TableRuntime runtime);
```
`TableConfiguration` will be serialized and persisted to sysdb

Reply:
ok

Reply:
When I first saw it, I also wondered if it could be unioned with `ExpiringDataConfig`.
Can we mark those parts that are not suitable for persistence as transient and reinitialize them when needed?